pkg/compact/schedule_delete.go

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package compact
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+// DebugCompactorMetas is folder to store compactor metadata
+// about when the block is scheduled to be deleted.
+const DebugCompactorMetas = "debug/compactor-metas"
+
+// CompactorMeta stores block id and when block was marked for deletion.
+type CompactorMeta struct {
+	ID           ulid.ULID `json:"id"`
+	DeletionTime int64     `json:"deletion_time"`
+}
+
+// ScheduleBlockDelete marks the block to be deleted.
+type ScheduleBlockDelete struct {
+	dir         string
+	logger      log.Logger
+	deleteDelay time.Duration
+	bkt         objstore.Bucket
+}
+
+// NewScheduleBlockDelete creates a new ScheduleBlockDelete.
+func NewScheduleBlockDelete(logger log.Logger, dir string, bkt objstore.Bucket, deleteDelay time.Duration) *ScheduleBlockDelete {
+	return &ScheduleBlockDelete{
+		dir:         dir,
+		logger:      logger,
+		deleteDelay: deleteDelay,
+		bkt:         bkt,
+	}
+}
+
+// ScheduleDelete deletes blocks from bucket
+// deleteDelay duration after block is marked for deletion.
+func (s *ScheduleBlockDelete) ScheduleDelete(ctx context.Context) error {
+	compactorMetaPath := filepath.Join(s.dir, "download")
+	defer func() {
+		if err := os.RemoveAll(compactorMetaPath); err != nil {
+			level.Error(s.logger).Log("msg", "failed to remove compactor meta dir", compactorMetaPath, "err", err)
+		}
+	}()
+
+	if err := os.RemoveAll(compactorMetaPath); err != nil {
+		return errors.Wrap(err, "clean compaction compactor meta dir")
+	}
+
+	if err := objstore.DownloadDir(ctx, s.logger, s.bkt, DebugCompactorMetas, compactorMetaPath); err != nil {
+		return errors.Wrap(err, "downlad compactor-metas dir")
+	}
+
+	return filepath.Walk(compactorMetaPath, func(path string, fileInfo os.FileInfo, err error) error {
+		_, file := filepath.Split(path)
+		if _, err := ulid.Parse(strings.TrimSuffix(file, filepath.Ext(file))); err != nil {
+			compactorMetaBytes, err := ioutil.ReadFile(path)
+			if err != nil {
+				return errors.Wrap(err, "read compactor meta")
+			}
+
+			compactorMeta := CompactorMeta{}
+
+			if err := json.Unmarshal([]byte(compactorMetaBytes), &compactorMeta); err != nil {
+				return errors.Wrap(err, "unmarshal compactor meta")
+			}
+
+			if time.Now().Unix()-compactorMeta.DeletionTime > s.deleteDelay.Milliseconds() {
```

Comment:
There is mismatch between units here. Unix time is in seconds, but comparison is against milliseconds.

It would be better to use typed values:

```suggestion
			if time.Since(time.Unix(compactorMeta.DeletionTime, 0)) > s.deleteDelay {
```